cmd/acra-rotate/dbRotation.go

```diff
@@ -0,0 +1,188 @@
+		return nil, err
+	}
+	keypair := store.keys[idStr]
+	return keypair.oldPrivatekey, nil
+}
+
+// rotateDb execute selectQuery to fetch AcraStructs with related zone ids, decrypt with rotated zone keys and
+func rotateDb(selectQuery, updateQuery string, db *sql.DB, keystore keystore.KeyStore) bool {
+	rows, err := db.Query(selectQuery)
+	if err != nil {
+		log.WithError(err).Errorf("Can't fetch result")
+		return false
+	}
+	defer rows.Close()
+	columns, err := rows.Columns()
+	if err != nil {
+		log.WithError(err).Errorln("Can't fetch info about result columns")
+		return false
+	}
+	if len(columns) < 2 {
+		log.Errorln("Result has < 2 columns. Expected at least ZoneId and AcraStruct")
+		return false
+	}
+
+	keysStore := &keyRotatationStore{keystore: keystore, keys: make(map[string]*keyPair)}
+
+	row := make([]interface{}, len(columns))
+	rowPointers := make([]interface{}, len(columns))
+	// dynamically create list of pointers to interface to pass to row.Scan method to fetch all values
+	for i := 0; i < len(columns); i++ {
+		rowPointers[i] = &row[i]
+	}
+	var extraArgs []interface{}
+	acraStructIDIndex := len(columns) - 2 // last but one item
+	acraStructIndex := len(columns) - 1   // last item
+	for rows.Next() {
+		err = rows.Scan(rowPointers...)
+		if err != nil {
+			log.WithError(err).Errorln("Can't read row from result")
+			return false
+		}
+
+		acraStructID, ok := row[acraStructIDIndex].([]byte)
+		// check that acrastruct and id have correct types ([]byte)
+		if !ok {
+			log.Errorf("ClientId/ZoneId column has incorrect type (bytes expected, took %s)", reflect.TypeOf(row[acraStructIDIndex]))
+			return false
+		}
+
+		acraStruct, ok := row[acraStructIndex].([]uint8)
+		if !ok {
+			log.Errorf("AcraStruct column has incorrect type (bytes expected, took %s)", reflect.TypeOf(row[acraStructIndex]))
+			return false
+		}
+		logger := log.WithFields(log.Fields{"ZoneId": string(acraStructID)})
+		logger.Infof("Rotate AcraStruct with ZoneId=%s", string(acraStructID))
+
+		// rotate
+		privateKey, err := keysStore.getPrivateKey(acraStructID)
+		if err != nil {
+			logger.WithField("acrastruct", hex.EncodeToString(acraStruct)).WithError(err).Errorln("Can't get private key")
+			return false
+		}
+		publicKey, err := keysStore.getPublicKey(acraStructID)
+		if err != nil {
+			logger.WithField("acrastruct", hex.EncodeToString(acraStruct)).WithError(err).Errorln("Can't load public key")
+			return false
+		}
+		decrypted, err := base.DecryptAcrastruct(acraStruct, privateKey, acraStructID)
+		if err != nil {
+			logger.WithField("acrastruct", hex.EncodeToString(acraStruct)).WithError(err).Errorln("Can't decrypt AcraStruct")
+			return false
+		}
+
+		rotated, err := acrawriter.CreateAcrastruct(decrypted, publicKey, acraStructID)
+		if err != nil {
+			logger.WithField("acrastruct", hex.EncodeToString(acraStruct)).WithError(err).Errorln("Can't rotate data")
+			return false
+		}
+		if len(rowPointers) > 2 {
```

Comment:
shall we zeroing decrypted data and private key on this step?

Reply:
data yes, private key no, because can be used in next iterations if data was encrypted with this private key that after rotation doesn't exists, only in memory of program. added zeroing after execution and zeroing decrypted data after re-encrypting

Reply:
awesome, thank you!